akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -1059,95 +1070,85 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef) extends Actor with
-          // and minimum number of nodes have joined the cluster
-          if (upNumber == 0) {
-            // It is alright to use same upNumber as already used by a removed member, since the upNumber
-            // is only used for comparing age of current cluster members (Member.isOlderThan)
-            val youngest = localGossip.youngestMember
-            upNumber = 1 + (if (youngest.upNumber == Int.MaxValue) 0 else youngest.upNumber)
-          } else {
-            upNumber += 1
-          }
-          m.copyUp(upNumber)
-
-        case m if m.status == Leaving ⇒
-          // Move LEAVING => EXITING (once we have a convergence on LEAVING)
-          m copy (status = Exiting)
-      }
+    val removedExitingConfirmed = exitingConfirmed.filter { n ⇒
+      val member = latestGossip.member(n)
+      member.team == selfTeam && member.status == Exiting
     }
 
-    if (removedUnreachable.nonEmpty || removedExitingConfirmed.nonEmpty || changedMembers.nonEmpty) {
-      // handle changes
-
-      // replace changed members
-      val newMembers = changedMembers.union(localMembers).diff(removedUnreachable)
-        .filterNot(m ⇒ removedExitingConfirmed(m.uniqueAddress))
-
-      // removing REMOVED nodes from the `seen` table
-      val removed = removedUnreachable.map(_.uniqueAddress).union(removedExitingConfirmed)
-      val newSeen = localSeen diff removed
-      // removing REMOVED nodes from the `reachability` table
-      val newReachability = localOverview.reachability.remove(removed)
-      val newOverview = localOverview copy (seen = newSeen, reachability = newReachability)
-      // Clear the VectorClock when member is removed. The change made by the leader is stamped
-      // and will propagate as is if there are no other changes on other nodes.
-      // If other concurrent changes on other nodes (e.g. join) the pruning is also
-      // taken care of when receiving gossips.
-      val newVersion = removed.foldLeft(localGossip.version) { (v, node) ⇒
-        v.prune(VectorClock.Node(vclockName(node)))
+    val changedMembers = {
+      val enoughMembers: Boolean = isMinNrOfMembersFulfilled
+      def isJoiningToUp(m: Member): Boolean = (m.status == Joining || m.status == WeaklyUp) && enoughMembers
+
+      latestGossip.members collect {
+        var upNumber = 0
+
+        {
+          case m if m.team == selfTeam && isJoiningToUp(m) ⇒
+            // Move JOINING => UP (once all nodes have seen that this node is JOINING, i.e. we have a convergence)
+            // and minimum number of nodes have joined the cluster
+            if (upNumber == 0) {
+              // It is alright to use same upNumber as already used by a removed member, since the upNumber
+              // is only used for comparing age of current cluster members (Member.isOlderThan)
+              val youngest = latestGossip.youngestMember
+              upNumber = 1 + (if (youngest.upNumber == Int.MaxValue) 0 else youngest.upNumber)
```

Comment:
yes, this is fine, we try to order them globally just because it's simple, but for different teams there is a high risk that same upNumber will be used for two different members. Clarify in docs of `Member.isOlderThan` that only makes sense to compare members of the same team.